libyul/optimiser/FunctionSpecializer.cpp

```diff
@@ -0,0 +1,149 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/optimiser/FunctionSpecializer.h>
+
+#include <libyul/optimiser/ASTCopier.h>
+#include <libyul/optimiser/NameCollector.h>
+#include <libyul/optimiser/NameDispenser.h>
+
+#include <libyul/AST.h>
+#include <libyul/YulString.h>
+#include <libsolutil/CommonData.h>
+
+#include <boost/algorithm/cxx11/any_of.hpp>
+
+#include <variant>
+
+using namespace std;
+using namespace solidity::util;
+using namespace solidity::yul;
+
+FunctionSpecializer::LiteralArguments FunctionSpecializer::specializableArguments(
+	FunctionCall const& _f
+)
+{
+	auto heuristic = [&](Expression const& _e) -> optional<Expression>
+	{
+		if (holds_alternative<Literal>(_e))
+			return ASTCopier{}.translate(_e);
+		return nullopt;
+	};
+
+	return applyMap(_f.arguments, heuristic);
+}
+
+void FunctionSpecializer::operator()(FunctionCall& _f)
+{
+	ASTModifier::operator()(_f);
+
+	if (m_dialect.builtin(_f.functionName.name))
+		return;
+
+	LiteralArguments arguments = specializableArguments(_f);
+
+	if (boost::algorithm::any_of(arguments, [](auto& _a) { return _a.has_value(); }))
+	{
+		auto const& oldName = _f.functionName.name;
```

Comment:
Could `move` here, since you're reassigning to `_f.functionName.name` later

Reply:
Done.